### DISCUSSION TITLE
2.0.0-beta1

### DISCUSSION THREAD

**Discussion by mnapoli** [brefphp/bref#1357] [General]
This is the first beta version of Bref v2.0 🎉

And we celebrate reaching **10 billion** Lambda invocations powered by Bref every month.
That's [1 in every 1000 AWS Lambda invocations that runs with Bref](https://twitter.com/matthieunapoli/status/1603032544424894464)!

OK, now with Bref v2. This first beta contains most of the changes for v2, even though I want to squeeze in more improvements before the stable release (especially improve the docs and the website, and the Laravel integration). In the meantime, the [Bref v2 roadmap is here](https://github.com/orgs/brefphp/projects/1/views/2) (but it is still open for ideas).

How to get started? Update `composer.json`:

```diff
-    "bref/bref": "^1.0",
+    "bref/bref": "^2.0@beta",
```

Then run `composer update bref/bref --with-all-dependencies`.

> **Note**
> Below is a short summary. **Read [Upgrading to v2](https://github.com/brefphp/bref/blob/v2/docs/upgrading/v2.md)** for more details.

How is v2 better than v1:

- The `bref/bref` package is lighter (fewer dependencies), which leaves more room for your code on Lambda and improves cold starts (15% faster).
- [Beta ARM (Graviton) runtimes](https://github.com/brefphp/bref/tree/v2/docs/runtimes#arm-runtimes)!
- Most `vendor/bin/bref` CLI commands have been replaced by much better and simpler `serverless bref:xxx` commands (https://github.com/brefphp/bref/pull/1303 provides an overview, check out the [v2 docs as well](https://github.com/brefphp/bref/tree/v2/docs)).
- Runtimes have been rewritten from scratch to use Red Hat's very popular [Remi's RPM repository](https://rpms.remirepo.net/), leading to cleaner internals, more reliable PHP binaries and builds, and much faster builds! No direct benefit for users except the confidence that Bref is in a much better position to last a long time :)
- Bref internals have been reworked, improved and automated a lot, leading to a more contributor-friendly codebase, with easier maintenance.
- Bref is now easier to use with AWS CDK, Pulumi, or other Node tools thanks to [the new `@bref.sh/layers` NPM package](https://github.com/brefphp/layers.js) that exposes layer versions. (hint: I have more ambitious ideas on this, this is just the start)

Overview of breaking changes:

- PHP 8.0 or greater is required
- The `serverless` CLI v3 is required (run `serverless --version` to check)
- `vendor/bin/bref cli` is replaced by the simpler `serverless bref:cli`
- `vendor/bin/bref local` is replaced by the simpler `serverless bref:local`
- `vendor/bin/bref layers` is replaced by the simpler `serverless layers`
- Docker images for local dev have been improved and simplified: `bref/fpm-dev-gateway` is gone, you need to update `docker-compose.yml` if you have one
- The `separateVendor` option in `serverless.yml` has been removed

> **Note**
> Again, read [Upgrading to v2](https://github.com/brefphp/bref/blob/v2/docs/upgrading/v2.md) for instructions on dealing with each of those changes.

Note that Bref Extra extensions are in the process of being updated for v2, you can follow https://github.com/brefphp/extra-php-extensions/pull/377 

A huge thank you to all contributors that have helped to maintain v1 and/or with this huge upcoming release, including @deleugpn who rewrote most of the Lambda runtimes, @GrahamCampbell who is helping to keep the v1 runtimes up to date, @t-richard and @afu-dev who have been extremely helpful on the Symfony integration and in the community, @Nyholm who has been maintaining flawlessly the Bref Extra extensions, @mykiwi for keeping PHP up to date, @shouze who is helping on layers and extra extensions, @georgeboot who is contributing amazing ideas on the Laravel integrations, @tillkruss who is pushing what Bref could be for Laravel users, and so many more (@shadowhand, @kevincerro…).

---

Finally, a few things I wanted to share:

- Bref has a Twitter account: https://twitter.com/brefphp
- Bref also has an awesome Slack community: https://bref.sh/docs/community.html
- I recently launched [Bref Pro and Bref Enterprise](https://bref.sh/#plans): get professional support while helping me work full time on Bref ❤️

And a huge thank you to all Bref sponsors, which [now includes AWS as well](https://twitter.com/matthieunapoli/status/1602339013490315264) 🎉 

<img width="727" alt="image" src="https://user-images.githubusercontent.com/720328/207851832-a6915791-0d24-441f-9fc4-7e28f94ccd9a.png">

<hr /><em>This discussion was created from the release <a href='https://github.com/brefphp/bref/releases/tag/2.0.0-beta1'>2.0.0-beta1</a>.</em>

**Comment by aran112000** [comment]
Amazing work @mnapoli, super keen to be able to get using the ARM layers, do you have a loose timeline for PHP 8.1 support on ARM?

**Reply by mnapoli** [reply]
Unfortunately I don't because we are tied to Amazon Linux Extra packages for ARM, and it doesn't support 8.1.

That being said, I followed https://repost.aws/questions/QUsfhDbww4SAy8i5Jmd4vcMg/amazon-linux-2-amazon-linux-extras-php-8-1-support which lead me to https://github.com/aws/elastic-beanstalk-roadmap/projects/1 which lead me to https://github.com/aws/elastic-beanstalk-roadmap/issues/214#issuecomment-1347213681 and this is what I read, posted **3 days ago**:

> So it seems like we now have PHP 8.1.x via amazon-linux-extras

I need to check now!!

**Reply by mnapoli** [reply]
PHP 8.1 is now supported with ARM layers 🎉

https://github.com/brefphp/bref/releases/tag/2.0.0-beta2

**Reply by shadowhand** [reply]
Does this mean that we might have an ARM version of the Bref dev container soon? 🤞🏼

**Reply by mnapoli** [reply]
We do have them: https://hub.docker.com/r/bref/arm-php-81-fpm-dev Or did you mean something else?

**Reply by shadowhand** [reply]
Oh I missed that! Awesome!

**Comment by aran112000** [comment]
I did a check earlier before posting (sorry should have said), there is a enablable repo in amazon-linux-extras for 8.1, yeah :)

**Reply by mnapoli** [reply]
It seems to work! I'm working on a PR 👍

**Comment by theofidry** [comment]
Looking forward to try it out!

**Comment by ctessmer** [comment]
@mnapoli This is an excellent write up! It persuaded me to try it out. Thanks for all your hard work! (And very cool how you give thanks to everyone else; the mark of a superb leader.)

**Reply by mnapoli** [reply]
Thank you for the feedback @ctessmer!
